File: apps/api/app/schemas/menu.py
import uuid

from pydantic import BaseModel


class ModifierOptionOut(BaseModel):
    id: uuid.UUID
    name: str
    price_cents: int
    is_default: bool
    sort_order: int

    model_config = {"from_attributes": True}


class ModifierGroupOut(BaseModel):
    id: uuid.UUID
    name: str
    description: str | None = None
    min_select: int
    max_select: int
    is_required: bool
    sort_order: int
    options: list[ModifierOptionOut]

    model_config = {"from_attributes": True}


class MenuItemOut(BaseModel):
    id: uuid.UUID
    name: str
    description: str | None = None
    price_cents: int
    sort_order: int
    is_active: bool
    modifier_groups: list[ModifierGroupOut]

    model_config = {"from_attributes": True}


class MenuCategoryOut(BaseModel):
    id: uuid.UUID
    name: str
    description: str | None = None
    sort_order: int
    is_active: bool
    items: list[MenuItemOut]

    model_config = {"from_attributes": True}


class MenuOut(BaseModel):
    id: uuid.UUID
    name: str
    is_active: bool
    categories: list[MenuCategoryOut]

    model_config = {"from_attributes": True}


class MenuItemIn(BaseModel):
    name: str
    description: str | None = None
    price_cents: int
    sort_order: int | None = None
    is_active: bool = True


class MenuCategoryIn(BaseModel):
    name: str
    description: str | None = None
    sort_order: int | None = None
    is_active: bool = True
    items: list[MenuItemIn]


class MenuUpsert(BaseModel):
    name: str
    is_active: bool = True
    categories: list[MenuCategoryIn]

File: apps/web/src/lib/menu.ts
import type { MenuOut } from "./types";
import type { MenuCategory } from "./restaurantDemo";

export function menuFromApi(menu: MenuOut): MenuCategory[] {
  return menu.categories
    .filter((category) => category.is_active)
    .map((category) => ({
      name: category.name,
      items: category.items
        .filter((item) => item.is_active)
        .map((item) => ({
          id: item.id,
          name: item.name,
          description: item.description || "",
          price: item.price_cents / 100,
          modifierGroups: item.modifier_groups.map((group) => ({
            id: group.id,
            name: group.name,
            description: group.description || undefined,
            minSelect: group.min_select,
            maxSelect: group.max_select,
            isRequired: group.is_required,
            options: group.options.map((option) => ({
              id: option.id,
              name: option.name,
              price: option.price_cents / 100,
              isDefault: option.is_default,
            })),
          })),
        })),
    }))
    .filter((category) => category.items.length > 0);
}
